mcs/class/System/Microsoft.CSharp/CSharpCodeCompiler.cs

```diff
@@ -62,10 +62,20 @@ static CSharpCodeCompiler ()
 				MethodInfo get_gac = gac.GetGetMethod (true);
 				string p = Path.GetDirectoryName (
 					(string) get_gac.Invoke (null, null));
-				windowsMonoPath = Path.Combine (
-					Path.GetDirectoryName (
-						Path.GetDirectoryName (p)),
-					"bin\\mono.bat");
+
+				string processExe = Process.GetCurrentProcess ().MainModule.FileName;
+				if (processExe != null) {
+					string fileName = Path.GetFileName (processExe);
+					if (fileName.StartsWith ("mono") && (fileName.EndsWith (".exe") || fileName.EndsWith (".bat"))) {
+						windowsMonoPath = processExe;
+					}
+				}
+
+				if (windowsMonoPath == null || !File.Exists (windowsMonoPath))
+					windowsMonoPath = Path.Combine (
+						Path.GetDirectoryName (
+							Path.GetDirectoryName (p)),
+						"bin\\mono.bat");
```

Comment:
Mono nowadays doesn't ship with a mono.bat file anymore so we can kill this codepath.